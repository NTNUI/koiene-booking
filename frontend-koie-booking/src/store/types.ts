// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
