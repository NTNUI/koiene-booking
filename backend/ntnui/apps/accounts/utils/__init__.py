from accounts.utils.mail_utils import send_gsuite_confirmation_mail
