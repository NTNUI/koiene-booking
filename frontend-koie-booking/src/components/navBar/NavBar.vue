<template>
  <v-app-bar app color="backgroundLight" dark>
    <div>
      <v-btn target="_blank" text @click="goToAllKoier">
        <h1 class="headline">NTNUI Koiene</h1>
      </v-btn>
    </div>
    <v-spacer />
    <NavBarLogin />
  </v-app-bar>
</template>

<script lang="ts">
import Vue from 'vue';
import NavBarLogin from './NavBarLogin.vue';

export default Vue.extend({
  name: 'App',
  components: {
    NavBarLogin
  },
  methods: {
    goToAllKoier() {
      if (this.$router.currentRoute.path !== '/') {
        this.$router.push(`/`);
      }
    }
  }
});
</script>

<style></style>
