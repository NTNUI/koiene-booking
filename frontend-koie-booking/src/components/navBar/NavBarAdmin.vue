<template ref="navBarAdmin">
  <v-btn id="navBarAdminButton" text @click="$router.push('/admin')">
    Admin
  </v-btn>
</template>

<script>
export default {
  name: 'NavBarAdmin'
};
</script>

<style scoped></style>
