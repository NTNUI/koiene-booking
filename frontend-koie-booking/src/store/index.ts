import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
