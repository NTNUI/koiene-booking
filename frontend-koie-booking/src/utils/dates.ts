import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
