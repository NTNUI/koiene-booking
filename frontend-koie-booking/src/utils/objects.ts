export const clone = (object: any) => Object.assign({}, object);

export const renameKey = (object: any, key: string, newKey: string): any => {
  const value = object[key];
  delete object[key];
  object[newKey] = value;
  return object;
};

export const sortObjectByKey = (unordered: any) => {
  let ordered: any;
  ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key) {
      ordered[key] = unordered[key];
    });
  return ordered;
};
