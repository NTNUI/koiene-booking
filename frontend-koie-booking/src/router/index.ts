import Vue from 'vue';
import VueRouter from 'vue-router';
import store from '@/store';

Vue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'AllKoier',
    component: () => import(/* webpackChunkName: "about" */ '../views/AllKoier.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/koie/:id',
    name: 'koie',
    component: () => import('../views/Koie.vue'),
    props: true,
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/booking/:id',
    name: 'booking',
    component: () => import('../views/Booking.vue'),
    meta: { requiresAuth: true, requiresAdmin: false }
  },
  {
    path: '/login',
    name: 'Login',
    component: () => import('../views/Login.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/admin',
    name: 'Admin',
    component: () => import('../views/Admin.vue'),
    meta: { requiresAuth: true, requiresAdmin: true }
  }
];

const router = new VueRouter({
  routes
});

export function beforeEach(to: any, from: any, next: any) {
  const isLoggedIn = store.getters['auth/isLoggedIn'];
  const isAdmin = store.getters['auth/isAdmin'];
  const requiresAuth = to.matched.some((record: any) => record.meta.requiresAuth);
  const requiresAdmin = to.matched.some((record: any) => record.meta.requiresAdmin);

  if (requiresAdmin && !isAdmin) next({ name: 'Login', query: { redirect: to.path } });
  else if (requiresAuth && !isLoggedIn) next({ name: 'Login', query: { redirect: to.path } });
  else if (to.name === 'Login' && isLoggedIn) next('/');
  else next();
}

router.beforeEach((to, from, next) => beforeEach(to, from, next));

export default router;
export { routes };
