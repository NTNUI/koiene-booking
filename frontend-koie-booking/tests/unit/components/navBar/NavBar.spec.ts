import Vue from 'vue';
import Vuetify from 'vuetify';
import Vuex from 'vuex';
import { storeConfig } from '@/store';
import { cloneDeep } from 'lodash';

Vue.use(Vuetify);

// Components or views

import NavBar from '@/components/navBar/NavBar.vue';
import NavBarAdmin from '@/components/navBar/NavBarAdmin.vue';
import { createWrapper } from '../../utils';
import VueRouter from 'vue-router';
import { routes } from '@/router';
const router = new VueRouter({ routes, mode: 'abstract' });

describe('Component NavBar.vue', () => {
  it('Matches snapshot', async () => {
    const wrapper = createWrapper(NavBar, { router });

    expect(wrapper).toMatchSnapshot();
  });

  it('Admin button pushes the router to /admin', () => {
    // Arrange
    const store = new Vuex.Store(cloneDeep(storeConfig));
    store.commit('auth/setAdmin', true);
    store.commit('auth/setToken', { access: 'testToken' });
    const wrapper = createWrapper(NavBar, {
      store,
      router,
    });
    const spy = jest.spyOn(wrapper.vm.$router, 'push');

    // Act
    wrapper.find('#navBarAdminButton').trigger('click');

    // Assert
    expect(spy).toHaveBeenCalledWith('/admin');
  });

  it('Admin can see admin button if logged in', () => {
    // Arrange
    const store = new Vuex.Store(cloneDeep(storeConfig));

    store.commit('auth/setAdmin', true);
    store.commit('auth/setToken', { access: 'testToken' });

    // Act
    const wrapper = createWrapper(NavBar, { store, router });

    //Assert
    expect(wrapper.contains(NavBarAdmin)).toBe(true);
  });

  it("User can't see admin button if not admin but logged in", () => {
    // Arrange
    const store = new Vuex.Store(cloneDeep(storeConfig));

    store.commit('auth/setToken', { access: 'testToken' });
    store.commit('auth/setAdmin', false);

    // Act
    const wrapper = createWrapper(NavBar, { store, router });

    // Assert
    expect(wrapper.contains(NavBarAdmin)).toBe(false);
  });

  it("Admin can't see admin button if not logged in", () => {
    // Arrange
    const store = new Vuex.Store(cloneDeep(storeConfig));
    store.commit('auth/setAdmin', true);

    // Act
    const wrapper = createWrapper(NavBar, {
      store,
      router,
    });

    // Assert
    expect(wrapper.contains(NavBarAdmin)).toBe(false);
  });
});
