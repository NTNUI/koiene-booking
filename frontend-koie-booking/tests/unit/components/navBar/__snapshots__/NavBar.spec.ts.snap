// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component NavBar.vue Matches snapshot 1`] = `
<header class="v-sheet theme--dark v-toolbar v-app-bar v-app-bar--fixed backgroundLight" style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;">
  <div class="v-toolbar__content" style="height: 64px;">
    <div><button target="_blank" type="button" class="v-btn v-btn--flat v-btn--text theme--dark v-size--default"><span class="v-btn__content"><h1 class="headline">NTNUI Koiene</h1></span></button></div>
    <div class="spacer"></div>
    <!----> <button type="button" class="mr-4 v-btn v-btn--contained theme--dark v-size--default"><span class="v-btn__content">
  LOG IN
</span></button>
  </div>
</header>
`;
